consumer/src/main/java/com/flipkart/varadhi/consumer/GroupedMessageSrc.java

```diff
@@ -47,22 +66,65 @@ public class GroupedMessageSrc<O extends Offset> implements MessageSrc {
      */
     @Override
     public CompletableFuture<Integer> nextMessages(MessageTracker[] messages) {
-        if (!hasMaxInFlightMessages()) {
-            return replenishAvailableGroups().thenApply(v -> nextMessagesInternal(messages));
+        int count = nextMessagesInternal(messages);
+        if (count > 0) {
+            return CompletableFuture.completedFuture(count);
+        }
+
+        NextMsgsRequest request = new NextMsgsRequest(new CompletableFuture<>(), messages);
+        if (!pendingRequest.compareAndSet(null, request)) {
+            throw new IllegalStateException(
+                    "nextMessages method is not supposed to be called concurrently. There seems to be a pending nextMessage call");
+        }
+
+        // incomplete result is saved. trigger new message fetch.
+        optionallyFetchNewMessages();
+
+        // double check, if any free group is available now.
+        if (isFreeGroupPresent()) {
+            tryCompletePendingRequest();
+        }
+
+        return request.result;
+    }
+
+    private void tryCompletePendingRequest() {
+        NextMsgsRequest request;
+        if ((request = pendingRequest.getAndSet(null)) != null) {
+            request.result.complete(nextMessagesInternal(request.messages));
```

Comment:
there is still a chance, where we still end up completing the request with 0 messages, but this should be much rarer now.